### PR TITLE
FVDE (FileVault 2) OpenCL support

### PR DESCRIPTION
**Pull request by kholia**
The format has "XXX is this correct?" in two places. I am not sure if I am doing the right thing at these places. Please review.

### CODE REVIEW

File: src/opencl_fvde_fmt_plug.c [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+	HANDLE_CLERROR(cl_error, _string);
+
+#define CLKERNELARG(kernel, id, arg, msg)\
+	HANDLE_CLERROR(clSetKernelArg(kernel, id, sizeof(arg), &arg), msg);
+
+	host_pass = mem_calloc(kpc, sizeof(pass_t));
+	host_crack = mem_calloc(kpc, sizeof(crack_t));
+	host_salt = mem_calloc(1, sizeof(salt_t));
+	cracked = mem_calloc(kpc, sizeof(*cracked));
+	cracked_size = kpc; // XXX is this correct?
```
Comment (magnumripper) [inline src/opencl_fvde_fmt_plug.c]: This should be `cracked_size = kpc * sizeof(*cracked_size);` (or perhaps `cracked_size = kpc * sizeof(int);`)

Comment (magnumripper) [inline src/opencl_fvde_fmt_plug.c]: Actually I think `sizeof(int)` is easier to "follow". But it's obviously the same thing.

Comment (magnumripper) [inline src/opencl_fvde_fmt_plug.c]: Hmmm and perhaps even like this
```
	cracked_size = kpc * sizeof(*cracked);
	cracked = mem_calloc(cracked_size, 1);
```

Comment (kholia) [inline src/opencl_fvde_fmt_plug.c]: I went with the last option. CI testing is in progress now. Thanks!

**Comment by magnumripper** [comment]
Travis still crashes. Not sure what's wrong. I see you fail to free `cracked` in `release_clobj()` but that's not the problem.

**Comment by kholia** [comment]
There must be a bug somewhere but I can't find it. I tested this on local GTX 960, `super` box and on Intel OpenCL too.

**Comment by kholia** [comment]
https://travis-ci.org/kholia/JohnTheRipper/builds/204618628 shows almost the same code as passing, weird!

**Comment by kholia** [comment]
I tried building JtR jumbo with ASan on my local Ubuntu 16.10 box which has a GTX 960. The build succeeds but `../run/john --list=opencl-devices` finds no OpenCL devices. Non-ASan builds on the same box run OpenCL formats just fine!

I have faced this problem before too, and I don't know the solution. Overall, my idea of using ASan to debug this problem failed.

**Comment by magnumripper** [comment]
Yeah ASan seems to hide nvidia GPUs for some reason. You could install fglrx or something like that to get AMD's CPU device just like Travis CI does it.

**Comment by magnumripper** [comment]
Tried with ASan on OSX. Tried all OpenCL devices using --test, --test=0 and --test-full=0. No problem with anything.

This might be a bug in the AMD CPU runtime Travis uses. However, we run the same kernel with RAR5-opencl and PBKDF2-HMAC-SHA256-opencl and they work fine on Travis.

**Comment by magnumripper** [comment]
> However, we run the same kernel with RAR5-opencl and PBKDF2-HMAC-SHA256-opencl and they work fine on Travis.

...so I just opened RAR5-opencl and FVDE-opencl side by side in Meld and stared at them for a while. I see nothing wrong.

**Comment by magnumripper** [comment]
I think we should just implement fvde_decrypt() on GPU and see if the problem goes away in the process. Will you do it or should I?

**Comment by kholia** [comment]
Instead of trying to fix and merge this code, getting pure GPU `FVDE-OpenCL` format implemented and merged sounds better!

You go ahead. This format might be too tricky for me :)

**Comment by magnumripper** [comment]
Okay. Do I have write access to your repo? If not, please give me that so I can just add to this PR.

**Comment by kholia** [comment]
Cool. I have added you as a collaborator now.

**Comment by kholia** [comment]
The build fails with almost the same ASan stack trace! Travis CI doesn't allow SSH access for unpaid plans, so debugging this is going to be hard.

Should we ignore the bot results for this particular ASan + OpenCL combination? Is this a sane idea?

**Comment by claudioandre** [comment]
I would say it is not behaving as a regular OpenCL format (I will comment on a review later).

On Travis:
```
Testing: FVDE-OpenCL, FileVault 2 [PBKDF2-SHA256 OpenCL AES]... modprobe: ERROR: ../libkmod/libkmod-module.c:809 kmod_module_insert_module() could not find module by name='fglrx'
modprobe: ERROR: could not insert 'fglrx': Function not implemented
Error: Fail to load fglrx kernel module!
Error! Fail to load fglrx kernel module! Maybe you can switch to root user to load kernel module directly
ASAN:SIGSEGV
=================================================================
==14399== ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x7f5699674e02 sp 0x7ffcc56f8080 bp 0x8daa90c38fc66084 T0)
AddressSanitizer can not provide additional info.
    #0 0x7f5699674e01 (/usr/lib/x86_64-linux-gnu/libasan.so.0+0x7e01)
    #1 0x7f5699682367 in __interceptor_free (/usr/lib/x86_64-linux-gnu/libasan.so.0+0x15367)
    #2 0x7f56946c446c (/usr/lib/libatiadlxx.so+0x2146c)
    #3 0x7f56946d8bd4 in ADL2_Main_Control_Destroy (/usr/lib/libatiadlxx.so+0x35bd4)
    #4 0x7f56885ac182 (/usr/lib/libamdocl64.so+0x679182)
    #5 0x7f56885ac2fb (/usr/lib/libamdocl64.so+0x6792fb)
    #6 0x7f56885ac723 (/usr/lib/libamdocl64.so+0x679723)
    #7 0x7f56885aed7a (/usr/lib/libamdocl64.so+0x67bd7a)
    #8 0x7f56885c8e16 (/usr/lib/libamdocl64.so+0x695e16)
    #9 0x7f5688599992 in clIcdGetPlatformIDsKHR (/usr/lib/libamdocl64.so+0x666992)
    #10 0x7f5698bb96ed (/usr/lib/libOpenCL.so.1+0x26ed)
    #11 0x7f5698bbb536 (/usr/lib/libOpenCL.so.1+0x4536)
    #12 0x7f5697c2ba7f in __GI___pthread_once (/lib/x86_64-linux-gnu/libpthread.so.0+0xda7f)
    #13 0x7f5698bbb5c6 (/usr/lib/libOpenCL.so.1+0x45c6)
    #14 0x7f5698bb9684 (/usr/lib/libOpenCL.so.1+0x2684)
    #15 0x7f5698bb9c80 in clGetPlatformIDs (/usr/lib/libOpenCL.so.1+0x2c80)
    #16 0x99a577 in start_opencl_environment /home/travis/build/claudioandre/JohnTheRipper/src/common-opencl.c:428
    #17 0x9a0939 in opencl_preinit /home/travis/build/claudioandre/JohnTheRipper/src/common-opencl.c:702
    #18 0x9a0dd9 in opencl_prepare_dev /home/travis/build/claudioandre/JohnTheRipper/src/common-opencl.c:2036
    #19 0x76b0d8 in init /home/travis/build/claudioandre/JohnTheRipper/src/opencl_fvde_fmt_plug.c:166
    #20 0x9061d7 in fmt_init /home/travis/build/claudioandre/JohnTheRipper/src/formats.c:81
    #21 0x924ecc in ldr_init_test_db /home/travis/build/claudioandre/JohnTheRipper/src/loader.c:1235
    #22 0x8ebdf2 in benchmark_all /home/travis/build/claudioandre/JohnTheRipper/src/bench.c:806
    #23 0x918b92 in john_run /home/travis/build/claudioandre/JohnTheRipper/src/john.c:1574
    #24 0x91a1f5 in main /home/travis/build/claudioandre/JohnTheRipper/src/john.c:1955
    #25 0x7f569787af44 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21f44)
    #26 0x406a72 in _start (/home/travis/build/claudioandre/JohnTheRipper/run/john+0x406a72)
SUMMARY: AddressSanitizer: SEGV ??:0 ??
==14399== ABORTING
```

PS: it is possible to get proper instrumented code for older gcc. But my patch can't be applied clean on master.

File: src/opencl_fvde_fmt_plug.c [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+#include "misc.h"
+#include "arch.h"
+#include "fvde_common.h"
+#include "common.h"
+#include "stdint.h"
+#include "formats.h"
+#include "options.h"
+#include "common-opencl.h"
+
+#define FORMAT_LABEL            "FVDE-OpenCL"
```
Comment (claudioandre) [inline src/opencl_fvde_fmt_plug.c]: You should change this to "FVDE-opencl". At least, the format 'filter' is not working properly.

**Comment by claudioandre** [comment]
Nodoby else has hardware to debug this? :sob: Wow, bad news indeed. Debug on CI is not a nice or easy thing to do. :scream: 

BTW: I haven't tried it on super (since I guess it won't show the expected behavior).

**Comment by magnumripper** [comment]
Super has the AMD CPU driver, but can't build with ASan.

**Comment by magnumripper** [comment]
```
    #16 0x99a577 in start_opencl_environment /home/travis/build/claudioandre/JohnTheRipper/src/common-opencl.c:428
    #17 0x9a0939 in opencl_preinit /home/travis/build/claudioandre/JohnTheRipper/src/common-opencl.c:702
    #18 0x9a0dd9 in opencl_prepare_dev /home/travis/build/claudioandre/JohnTheRipper/src/common-opencl.c:2036
    #19 0x76b0d8 in init /home/travis/build/claudioandre/JohnTheRipper/src/opencl_fvde_fmt_plug.c:166
```
This makes no sense to me. The RAR5 format is identical, doesn't crash.

```
   163	static void init(struct fmt_main *_self)
   164	{
   165		self = _self;
   166		opencl_prepare_dev(gpu_id);
   167	}
```

**Comment by solardiz** [comment]
On Tue, Feb 28, 2017 at 10:35:57AM -0800, magnum wrote:
> Super has the AMD CPU driver, but can't build with ASan.

Hmm, not even after "scl enable devtoolset-3 bash"?  Why not?

```
$ rpm -qa | fgrep asan
libasan-4.9.1-10.el6.i686
libasan-4.9.1-10.el6.x86_64
devtoolset-2-libasan-devel-4.8.2-15.1.el6.i686
devtoolset-3-libasan-devel-4.9.1-10.el6.x86_64
devtoolset-2-libasan-devel-4.8.2-15.1.el6.x86_64
```

**Comment by claudioandre** [comment]
> This makes no sense to me. The RAR5 format is identical, doesn't crash.

No, they were not. Now they are, or can be. 

At least, one build is over now, and it works [1]
```
Testing: enpass-opencl, Enpass Password Manager [PBKDF2-SHA1 AES OpenCL]... PASS
Testing: FVDE-opencl, FileVault 2 [PBKDF2-SHA256 OpenCL AES]... PASS
Testing: iwork-opencl, Apple iWork '09 / '13 / '14 [PBKDF2-SHA1 AES OpenCL]... PASS
```

[1] S3 is drunk (S3 outage in AWS), so you guys should rebuild it later.

**Comment by claudioandre** [comment]
Updating, it succeeded in two different builds now. 

BTW: we need a new CPU format
```
All 666 formats passed self-tests!
```

**Comment by kholia** [comment]
I never imagined that changing `FVDE-OpenCL` to `FVDE-opencl` would fix the problem! I don't understand the underlying problem. Nevertheless, good detective work @claudioandre. Also, it is nice to have resolved ASan stack traces.

**Comment by magnumripper** [comment]
>> Super has the AMD CPU driver, but can't build with ASan.

> Hmm, not even after "scl enable devtoolset-3 bash"?  Why not?

My bad, that works fine.

**Comment by magnumripper** [comment]
> I never imagined that changing FVDE-OpenCL to FVDE-opencl would fix the problem!

+1, especially since it worked just fine on OSX or without ASan. What the heck happened there? Maybe we need to fix some core code.

**Comment by magnumripper** [comment]
OK, all checks for the -opencl suffix in a label are using strstr, not strcasestr. So some required initializations for GPU were not done. Instead of changing that, I simply added a self-test check now that fails if the suffix isn't lower case. Thanks Claudio, you probably saved us DAYS 😆

**Comment by magnumripper** [comment]
> BTW: we need a new CPU format
```
All 666 formats passed self-tests!
```

We shouldn't count the 13th then. So it'll be 667.

**Pull request merged**